### PR TITLE
Change Path into a newtype.

### PR DESCRIPTION
**Pull request by paulp**
When I wrote Path I didn't know that a newtype against
a primitive type allowed you to use it as a receiver.
This removes a whole pointless layer of wrapping. Go go.

### CODE REVIEW

**Comment by paulp** [comment]
I realize this is diff conflict central, sorry about that.

**Comment by EECOLOR** [comment]
As you can see in the travis build you missed one.

**Comment by EECOLOR** [comment]
The conflicts are not that bad and I'm happy to solve them.

**Comment by EECOLOR** [comment]
I wonder if it's really better to use `x.Str()` vs `string(x)`. According to the specs regular type conversions are free. My guess is that the go compiler is not smart enough to eliminate the method call. I would go for `string(x)`.

File: suffuse/path.go [hunk 1]
```diff
@@ -39,47 +38,47 @@ func MaybeByteString(result []byte, err error) string {
...
-func (x Path) OsTruncate(size int64) error            { return os.Truncate(x.Path, size)        }
-
-func (x Path) OsCreate() (*os.File, error)                             { return os.Create(x.Path)               }
-func (x Path) OsLstat() (os.FileInfo, error)                           { return os.Lstat(x.Path)                }
-func (x Path) OsOpen() (*os.File, error)                               { return os.Open(x.Path)                 }
-func (x Path) OsOpenFile(flag int, perm os.FileMode) (*os.File, error) { return os.OpenFile(x.Path, flag, perm) }
-func (x Path) OsReadLink() (string, error)                             { return os.Readlink(x.Path)             }
-func (x Path) OsStat() (os.FileInfo, error)                            { return os.Stat(x.Path)                 }
+func (x Path) IsAbs() bool   { return filepath.IsAbs(x.Str()) }
+func (x Path) IsEmpty() bool { return string(x) == ""            }
```
Comment (EECOLOR) [inline suffuse/path.go:57]: If consistent, this should be `x.Str()`

**Comment by paulp** [comment]
The reason I didn't go with `string(x)` is that it's a cast, and I was thinking of the sort of problems shown here: https://gist.github.com/paulp/fca2ce1199abdfb26e69

But maybe that logic only applies with numeric types, where putting a cast into the source paves the way for future failure by abandoning all type safety. If I understand the go rules correctly those `string(x)` calls will stop compiling if we later changed Path to something else. In which case you're right, it should just be string(x).

**Comment by EECOLOR** [comment]
From the spec [Conversions](https://golang.org/ref/spec#Conversions):

> Specific rules apply to (non-constant) conversions between numeric types or to and from a string type. > These conversions may change the representation of x and incur a run-time cost. All other conversions only change the type but not the representation of x.

**Comment by EECOLOR** [comment]
Relevant part from the spec that relates to your example

> ... It is then truncated to fit in the result type's size. ... The conversion always yields a valid value ...

**Comment by paulp** [comment]
I'll fix this up.

**Comment by paulp** [comment]
@EECOLOR How about now.

**Comment by EECOLOR** [comment]
@paulp Looks good, time ran out for today, so I will merge it tomorrow. And after #20

**Comment by paulp** [comment]
I resolved the merge conflicts.

File: suffuse/path.go [hunk 2]
```diff
@@ -8,30 +8,23 @@ import (
   "io/ioutil"
 )
 
-type Path struct {
-  Path string
-}
-func NewPath(path string) Path {
-  return Path { path }
-}
-func NewPathRef(path string) *Path {
```
Comment (EECOLOR) [inline suffuse/path.go:17]: We need this one

File: suffuse/opts.go [hunk 3]
```diff
@@ -14,7 +14,7 @@ type sfsOpts struct {
 
 type SfsConfig struct {
   VolName string
-  Config *Path
+  Config Path
```
Comment (EECOLOR) [inline suffuse/opts.go:17]: Ahh, you changed this bit

File: suffuse/path.go (hunk 2 above)
Comment (EECOLOR) [inline suffuse/path.go:17]: Never mind

**Pull request merged**